### DISCUSSION TITLE
How to use this thing / guidance needed

### DISCUSSION THREAD

**Discussion by black-snow** [OP-TED/eForms-SDK#505] [Q&A] [answered]
This may sound like a silly question, but I don't get how to use the SDK. I scrolled through the docs, the sample apps, the java-core and toolkit, issues, and discussion but I still don't get it.

What I want to do is getting ye good olde POJOs from XMLs - in terms of this SDK this might be the "conceptual" representation from the "physical" (XML). I need to build applications atop all the different notice types and fields and ofc I don't want to read and understand every type and change things in my applications with every change to the XMLs. So I thought that the SDK was the way to go - to abstract away all the detail.

What I expected was:
* pull the SDK
* `SomeParsedDoc doc = SomeSdkXmlParser.parse("any-xml-less-or-equal-sdk-version.xml");`
* use the `doc` in my applicatoin logic

But it ain't seem to be any close to this. To me, it seems like the whole thing is made for creating and displaying documents and not to understand and process them. Did I miss a fundamental page in the docs?

![grafik](https://github.com/OP-TED/eForms-SDK/assets/579733/f9bea453-7e18-4ad1-9273-3d082edf9646)

**Comment by rouschr** [comment, marked as answer]
Hello,

The SDK provides the metadata to be used by any kind of application.
So yes those are just text files and apart from the EFX grammar, nothing close to code.
The xsds are the closest thing to the notice XML structure, but they will not help you for full validation (for that see schematron and CVS).

If you want to only load notices xml into POJOs the closest you will get is by using the schemas/maindoc/*.xsd files (4 files, one per doc type) and use something like JAXB or an equivalent. 
Once you get that to work, try loading files from /examples/notices/
I recommend you use the SDK 8 RC 1 as we fixed many issues: https://github.com/OP-TED/eForms-SDK/tree/1.8.0-rc.1

I suggest you have a look at the editor demo if you want to see SDK usage, but this is web based with a front/back end logic:
https://github.com/OP-TED/eforms-notice-editor

The editor demo is not covering everything right now but good enough to give you an idea on how to use the SDK.
There is a front end part (mostly HTML/CSS/Javascript) and a back end (Java).
The back end reads the SDK files from a folder and serves them as JSON to the front end to consume.
Some Javascript code reads JSON (fields.json + notice-types/xyz.json + codelist info, ...) and displays a form.
Once the form is filled the form data in transformed into JSON and sent to the server where it can be transformed to notice XML.

The SDK:
/fields/**fields.json** should be your main file if you are a developer as it links most things like fields, field containers (nodes), rules, codelists, ...
/notice-types folder proposes a way to display forms (creating new forms is covered by the editor demo)
/view-templates is about pure visualisation

And this part of the docs:
https://docs.ted.europa.eu/eforms/latest/guide/index.html
https://docs.ted.europa.eu/eforms/latest/guide/notice-forms.html
https://docs.ted.europa.eu/eforms/latest/guide/xml-generation.html

**Reply by black-snow** [reply]
Thanks for the quick reply @rouschr,

Could you post a rough draft how to tackle the problem? It's very overwhelming and hard to grasp from the outside. I peered into the `fields.json` and was blown away. Atm I have no idea what to do with it and the "schematrons" and "genericodes" and what have you.

It looks like I won't need schematrons at all (no need to validate).  
I wonder if the EFX stuff helps somehow. If they are used to visualize the notices they must be tied to the conceptual layer.  

I wish I could *just access* what's in Figure 1 here: https://docs.ted.europa.eu/eforms/latest/schema/notice-content.html  
Are there any plans to enhance the SDK to allow easy access to some sort of read model?

Another mildly related question: Why do you need to have several SDKs at once? Why can't you use, say the `1.8` SDK to parse everything that was generated with a version `<= 1.8`?

**Reply by rouschr** [reply]
Lots of questions at once.
I am not sure what your exact goal is, so I cannot really answer.
One thing is that the way things are shown for visualisation or in forms is not always matching the XML order or structure.
Things that repeat have to match, but everything else is free to vary.
The fields.json is the rosetta stone, which allows you to go from a form to an XML and the other way around.

> I wish I could just access what's in Figure 1 here:

For pure visualisation (just reading) you can use the **eforms notice viewer** project, the input is a notice xml and your desired language: https://github.com/OP-TED/eforms-notice-viewer or TED VIEWER.
Both are based on the SDK view-templates EFX files.
For example summary view:
```
// Summary
// sdkVersion='1.8.0-SNAPSHOT'
// File generated from metadata database version 1.7.100 created on the 2023-05-31T15:59:01
1 {ND-Root[OPT-300-Procedure-Buyer is present]} #{auxiliary|text|buyer} // 1 Buyer
	1 {ND-ContractingParty} #{auxiliary|text|buyer} // 1.1 Buyer
		{ND-ContractingParty} #{name|BT-500}:  ${for text:$orgid in OPT-300-Procedure-Buyer, text:$orgname in BT-500-Organization-Company[OPT-200-Organization-Company == $orgid] return  concat('[', $orgname, '] --- ')} // Official name
2 {ND-Root} #{auxiliary|text|procedure}  // 2. Procedure
	1 {ND-ProcedureProcurementScope} #{auxiliary|text|procedure} // 2.1 Procedure
...
```
It will show some generic information.
You could even come up with your own .efx templates if you have special needs, but you will have to maintain them yourself.

> Another mildly related question: Why do you need to have several SDKs at once? Why can't you use, say the 1.8 SDK to parse everything that was generated with a version <= 1.8?

EFX toolkit handles this the way you want, the viewer as well. They might be lagging a bit behind SDK releases at times.
But for the editor demo we will only support the latest version, also usually lagging behind the SDK.

Note that we are not forced to do something like the editor demo project, we provide this to be helpful and we use it ourselves to test some SDK features ideas before we release them. Actually the entire SDK is something done to be helpful ... as otherwise managing big regulation changes is hard for esenders (hard or just a lot of work) if the regulation changes. 
SDK versions allow to clearly define what kind of metadata change / structural changes applications have to handle, with release notes and so on, read more about SDK versioning here: https://docs.ted.europa.eu/eforms/latest/versioning.html

**Reply by black-snow** [reply]
@rouschr thanks. It's all about extracting the information from the XMLs without having to fully understand their structure or following in lock-step with every release or change in regulations. So, for example, I'd like to know who the buyer is given any notice (that contains one), i.e., his name, phone, email or whatever information is there. Or getting the description or place or awarding criteria, etc., and so further, and so on.

As for now, the SDK rather seems to be tailored towards publishing and not reading, is that correct? And if so, are there any plans to support _the other side_ as well?  
Is EFX the abstraction I'm looking for then? Or is it subject to change as much as the XMLs themselves are? I'll take a look at the toolkit.

I wonder if I said something offending. If the SDK is supposed to be helpful this should be the feedback to be sought after, shouldn't it? ;) It sure is a huge topic and a great SDK is exactly what's needed. That's why I'm here.

**Reply by black-snow** [reply]
MEMO: this might be the answer once it's available: https://docs.ted.europa.eu/rdf-conversion/ted-sws-introduction.html

**Reply by rouschr** [reply]
The fields.json of the SDK has fields, look for Buyer. Each field has an associated xpathAbsolute.
For phone numbers look for fields of type phone, for emails look for fields of type email.
You can use an xpath engine to evaluate the xpath, which should lead you to the desired notice XML elements.

**Reply by black-snow** [reply]
@rouschr will the XPath be the same for all notice types? Is there a (searchable) catalogue of what fields exist with a description of their contents?

**Reply by rouschr** [reply]
The xpath is the same for all notice types, but a field may not exist in all notice types.
You can find field xpathAbsolute here:
https://github.com/OP-TED/eForms-SDK/blob/1.8.0-rc.1/fields/fields.json
https://github.com/OP-TED/eForms-SDK/blob/1.7.0/fields/fields.json
...
